### PR TITLE
STORY-4557 - custom summary report grand totals row

### PR DESCRIPTION
**Pull request by albertovillalobos**
Sister branch to https://github.com/Invoca/web/pull/6085

### CODE REVIEW

**Comment by albertovillalobos** [comment]
Ready for CR @nburwell @Vicentecarrillo 
cc @mikeweaver

File: src/plugins/data_table/data_table.js [hunk 1]
```diff
@@ -243,6 +243,47 @@ var LocalDataTable = (function() {
...
+          this.dataTable.append("<tfoot><tr></tr></tfoot>");
+        }
+        // Iterate over the current columns config
+        var grandTotalsRowCellNodes = _.map(this.columnsConfig(), function(col) {
+          // If a column is visible, render its total
+          if (this.columnVisibility(col.id) || col.bulk) {
+            /***
+              TODO: Add some sort of warning if the user has a totals property BUT
+              doesn't have a isNontotalsColumn predicate test.
+            ***/
```
Comment (nburwell) [inline src/plugins/data_table/data_table.js]: do we need a warning?  why can't we allow a table where every single column has data?

Either way, please either add the warning or remove this TODO..

File: src/plugins/data_table/data_table.js [hunk 2]
```diff
@@ -243,6 +243,47 @@ var LocalDataTable = (function() {
...
+            if (this.isNontotalsColumn && this.isNontotalsColumn(col)) {
+              // If column is a non totals column, draw "Grand totals" on the first one and the rest are empty
+              if (hasGrandTotalsCell) {
+                return $("<td></td>");
+              } else {
+                hasGrandTotalsCell = true;
+                return $("<td>Grand totals</td>");
+              }
+            } else { // If the column is NOT a nonTotalsRow column
```
Comment (nburwell) [inline src/plugins/data_table/data_table.js]: please move this trailing comment to new line inside the block.. or actually isn't this line of comment needed? seems pretty obvious by the `if` check above..

File: src/plugins/data_table/data_table.js [hunk 3]
```diff
@@ -243,6 +243,47 @@ var LocalDataTable = (function() {
...
+              return node;
+            }
+          }
+        }, this);
+        // Clear footer
+        this.dataTable.find("tfoot tr").html("");
+        // Populate
+        grandTotalsRowCellNodes.forEach(function($td) {
+          this.dataTable.find("tfoot tr").append($td);
+        }.bind(this));
```
Comment (nburwell) [inline src/plugins/data_table/data_table.js]: is there a reason you build up an array of jquery objects, then clear the `tr`, then iterate across that array, versus this?

```
if (this.dataTable.find("tfoot").length < 1) {
  this.dataTable.append("<tfoot><tr></tr></tfoot>");
}

var footerRow = this.dataTable.find("tfoot tr");
footerRow.html("");

this.columnsConfig().forEach(function(col) {
  // mostly existing logic in the block
  footerRow.append(node);
});
```

That keeps it to one iteration and doesn't have the code broken up around finding/clearing rows.

File: src/plugins/data_table/server_side_data_table.js [hunk 4]
```diff
@@ -186,12 +186,21 @@ var ServerSideDataTable = (function() {
           // collection, so we must perform the check at this point as well.
           if (_.isUndefined(json.sEcho)) return;
           if (json.sEcho * 1 < oSettings.iDraw) return;
+          /***
+            TODO: Add some sort of warning if the user has a totals property BUT
+            doesn't have a isNontotalsColumn predicate test?
+          ***/
```
Comment (nburwell) [inline src/plugins/data_table/server_side_data_table.js]: this todo is redundant with one in data_table.js

File: examples_server/examples/data_table/server_side_with_totals_row/data_table.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    render : function() {
+      var collection  = new app.Collections.Books();
+      var table = new app.Views.BookTable({ collection : collection });
+      this.$el.html(table.render().$el);
+      return this;
+    }
+
+  });
+
+});
```
Comment (albertovillalobos) [inline examples_server/examples/data_table/server_side_with_totals_row/data_table.js]: Thanks

File: spec/plugins/data_table/server_side.js [hunk 6]
```diff
@@ -31,6 +31,30 @@ describe("DataTable Plugin", function() {
...
+          { name: 'Jon doe 3',  cost: '100',   type: 'boat',  description: "simple boat", resale_value: "50"},
+          { name: 'Jon doe 4',  cost: '100',   type: 'boat',  description: "simple boat", resale_value: "50"},
+          { name: 'Jon doe 5',  cost: '100',   type: 'boat',  description: "simple boat", resale_value: "50"},
+          { name: 'Jon doe 6',  cost: '100',   type: 'boat',  description: "simple boat", resale_value: "50"},
+          { name: 'Jon doe 7',  cost: '100',   type: 'boat',  description: "simple boat", resale_value: "50"},
+          { name: 'Jon doe 8',  cost: '100',   type: 'boat',  description: "simple boat", resale_value: "50"},
+          { name: 'Jon doe 9',  cost: '100',   type: 'boat',  description: "simple boat", resale_value: "50"},
+          { name: 'Jon doe 10', cost: '100',   type: 'boat',  description: "simple boat", resale_value: "50"}
+        ],
+        total: {name: null, cost: 1000, type: null, description: null, resale_value: 500 }
```
Comment (albertovillalobos) [inline spec/plugins/data_table/server_side.js]: Done

File: src/plugins/data_table/data_table.js (hunk 1 above)
Comment (albertovillalobos) [inline src/plugins/data_table/data_table.js]: I was thinking about this too, and didn't think it was worth it. TODO removed.

File: src/plugins/data_table/data_table.js (hunk 2 above)
Comment (albertovillalobos) [inline src/plugins/data_table/data_table.js]: Removed

File: src/plugins/data_table/data_table.js (hunk 3 above)
Comment (albertovillalobos) [inline src/plugins/data_table/data_table.js]: Done, much more efficient that way

File: src/plugins/data_table/server_side_data_table.js (hunk 4 above)
Comment (albertovillalobos) [inline src/plugins/data_table/server_side_data_table.js]: Removed

**Comment by albertovillalobos** [comment]
@nburwell Addressed 1st round of CR

**Comment by nburwell** [comment]
Looks good! 👍

**Comment by albertovillalobos** [comment]
Woohoo!
bitmoji

**Comment by mikeweaver** [comment]
LGTM

**Pull request merged**